NitroxClient/MonoBehaviours/MoonpoolManager.cs

```diff
@@ -0,0 +1,179 @@
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using NitroxClient.GameLogic;
+using NitroxClient.Unity.Helper;
+using NitroxModel.Core;
+using NitroxModel.DataStructures;
+using NitroxModel.DataStructures.GameLogic.Entities;
+using NitroxModel.DataStructures.GameLogic.Entities.Bases;
+using NitroxModel.DataStructures.Util;
+using NitroxModel_Subnautica.DataStructures;
+using UnityEngine;
+
+namespace NitroxClient.MonoBehaviours;
+
+/// <summary>
+/// Saves the moonpools NitroxEntity of a <see cref="Base"/> to assign them back after each <see cref="Base.RebuildGeometry"/>.
+/// </summary>
+/// <remarks>
+/// To recognize pieces even after the base is rebuilt, we use the base anchor (<see cref="Base.anchor"/>) to get an absolute cell value.
+/// </remarks>
+public class MoonpoolManager : MonoBehaviour
+{
+    private Entities entities;
+
+    private Base @base;
+    private NitroxId baseId;
+    private Dictionary<Int3, MoonpoolEntity> moonpoolsByCell;
+    private Dictionary<Int3, MoonpoolEntity> shiftedMoonpools;
+    public MoonpoolEntity LatestRegisteredMoonpool { get; private set; }
+
+    public void Awake()
+    {
+        entities = NitroxServiceLocator.LocateService<Entities>();
+
+        if (!TryGetComponent(out @base))
+        {
+            Log.Error($"Tried adding a {nameof(MoonpoolManager)} to a GameObject that isn't a bases, deleting it.");
+            Destroy(this);
+            return;
+        }
+        @base.TryGetNitroxId(out baseId);
+        moonpoolsByCell = new();
+        shiftedMoonpools = new();
+        @base.onPostRebuildGeometry += OnPostRebuildGeometry;
+    }
+
+    public void OnDestroy()
+    {
+        @base.onPostRebuildGeometry -= OnPostRebuildGeometry;
+    }
+    
+    public void LateAssignNitroxEntity(NitroxId baseId)
+    {
+        this.baseId = baseId;
+        foreach (MoonpoolEntity moonpoolEntity in moonpoolsByCell.Values)
+        {
+            moonpoolEntity.ParentId = baseId;
+            moonpoolEntity.Id = new(); // Generate a new id in case
```

Comment:
It will definitely be because this is only called when a moonpool is the first piece placed of a base, and the MoonpoolManager.LateAssignNitroxEntity instruction happens before the PlaceBase packet is sent